src/components/text-editor/prosemirror-adapter/plugins/link-plugin.ts

```diff
@@ -0,0 +1,214 @@
+
+export interface EditorLinkMenuEventDetail {
+    href: string;
+    text: string;
+}
+
+const updateLink = (
+    view: EditorView,
+    updateLinkCallback?: UpdateLinkCallback,
+) => {
+    const { from, to } = view.state.selection;
+
+    let text = '';
+    let href = '';
+    view.state.doc.nodesBetween(from, to, (node) => {
+        if (node.type.name === 'text') {
+            text = node.text;
+            node.marks.forEach((mark: Mark) => {
+                if (mark.type.name === 'link') {
+                    href = mark.attrs.href;
+                }
+            });
+        }
+    });
+
+    if (updateLinkCallback) {
+        updateLinkCallback(text, href);
+    }
+};
+
+export const createLinkPlugin = (updateLinkCallback?: UpdateLinkCallback) => {
+    return new Plugin({
+        key: linkPluginKey,
+        props: {
+            handlePaste: (view, event) => {
+                return processPasteEvent(view, event);
+            },
+            handleDOMEvents: {
+                mousedown: (view, event) => {
+                    if (
+                        (event.metaKey || event.ctrlKey) &&
+                        event.button === 0
+                    ) {
+                        return processModClickEvent(view, event);
+                    }
+
+                    return processClickEvent(view, event);
+                },
+            },
+        },
+        view: () => ({
+            update: (view) => {
+                updateLink(view, updateLinkCallback);
+            },
+        }),
+    });
+};
+
+const findStart = (doc, pos) => {
+    let start = pos;
+    while (start > 0) {
+        const before = doc.nodeAt(start - 1);
+        if (
+            before &&
+            before.isText &&
+            before.marks.some((mark) => mark.type.name === EditorMenuTypes.Link)
+        ) {
+            start--;
+        } else {
+            break;
+        }
+    }
+
+    return start;
+};
+
+const findEnd = (doc, pos) => {
+    let end = pos;
+    while (end < doc.content.size) {
+        const after = doc.nodeAt(end);
+        if (
+            after &&
+            after.isText &&
+            after.marks.some((mark) => mark.type.name === EditorMenuTypes.Link)
+        ) {
+            end++;
+        } else {
+            break;
+        }
+    }
+
+    return end;
+};
+
+const getLinkDataAtPosition = (view: EditorView, event: MouseEvent) => {
+    const pos = view.posAtCoords({ left: event.clientX, top: event.clientY });
+    if (pos) {
+        const node = view.state.doc.nodeAt(pos.pos);
+        if (
+            node &&
+            node.marks.some((mark) => mark.type.name === EditorMenuTypes.Link)
+        ) {
+            const linkMark = node.marks.find(
+                (mark) => mark.type.name === EditorMenuTypes.Link,
+            );
+            const href = linkMark.attrs.href;
+            const from = findStart(view.state.doc, pos.pos);
```

Comment:
For these two functions, could we use the `resolve` function to get more information on this `node`? (Because we are trying to find the start and end of the link node I believe)

Reply:
Is resolve a built in prosemirror function? I can look at that.